Automattic/jetpack
# Instagram Gallery: Improve Disconnection Flow

Part of #15551

#### Changes proposed in this Pull Request:


Introduce a `jetpack/instagram-gallery` Redux store to keep track of the current state of the Instagram access tokens used in a post editor.

- When an Instagram Gallery block is connected to Instagram, the access token is stored in state and marked as `connected`.
All blocks using that token will automatically know that the token is valid, and use it without double checking from the API.
- When an Instagram Gallery block is disconnected from Instagram, the access token in state is marked as `disconnected`.
All blocks using that token will automatically get disconnected as well, and their `accessToken` attribute cleared.
- This state does not persist, so on editor load, all existing Instagram Gallery blocks will need to check from the API before marking their token as `connected`, or `disconnected` (which would also clear the `accessToken` attribute).

#### Notes

This "sync" only works on disconnection.

Currently (before this PR even), if a user inserts a new IG block into a post containing another IG block connected to their own token, the new block will automatically connect.

Conversely (and this PR won't change the behaviour), if we insert two IG blocks in a post and connect the first one, the second won't connect automatically.
Reason is that new blocks don't have an `accessToken` set by default, so the second block doesn't know which token's status to check.

To be honest, I think this is not only a minor edge case, but also automatically connecting existing disconnected block might be a bit awkward and unexpected for the user.

#### Testing instructions:






Before checking out this PR, insert an IG block and disconnect it to start without tokens. Delete the block, save the post, and reload.

* Insert an IG block and connect it.
* Insert a second IG block: make sure it automatically connects.
* Disconnect one of the IG blocks.
* Make sure the other IG block gets disconnected as well.

More involved tests (starting without tokens again):

* Monitor the network for `wpcom/v2/instagram-gallery/access-token` requests.
* Insert an IG block: make sure it fires one `access-token` request.
* Connect it to Instagram.
* Insert another IG block: make sure it fires one `access-token` request as well, then it automatically connects to IG.
* **Duplicate** one of the connected IG blocks: make sure no `access-token` request is fired, and the block automatically connects to IG nonetheless.

#### Proposed changelog entry for your changes:

* N/A (unreleased block)
